File: android/app/src/main/java/dev/pam/nativeapp/modules/WatchChannel.kt
package dev.pam.nativeapp.modules

import java.util.ArrayDeque

internal class WatchChannel {
    private val queue = ArrayDeque<ByteArray>()
    private var waiter: ModuleCompletion? = null
    private var closed = false

    @Synchronized
    fun next(completion: ModuleCompletion) {
        if (closed) {
            completion.complete(ModuleResultStatus.FAILURE, "Observation is closed".toByteArray())
        } else if (queue.isNotEmpty()) {
            completion.complete(ModuleResultStatus.SUCCESS, queue.removeFirst())
        } else if (waiter != null) {
            completion.complete(ModuleResultStatus.FAILURE, "Observation already has a pending read".toByteArray())
        } else {
            waiter = completion
        }
    }

    fun offer(payload: ByteArray) {
        val callback = synchronized(this) {
            if (closed) return
            val value = waiter
            if (value == null) {
                if (queue.size >= 4) queue.removeFirst()
                queue.addLast(payload)
            } else {
                waiter = null
            }
            value
        }
        callback?.complete(ModuleResultStatus.SUCCESS, payload)
    }

    fun close() {
        val callback = synchronized(this) {
            if (closed) return
            closed = true
            queue.clear()
            val value = waiter
            waiter = null
            value
        }
        callback?.complete(ModuleResultStatus.FAILURE, "Observation stopped".toByteArray())
    }
}

File: ios/Sources/PamNative/Modules/WatchChannel.swift
import Foundation

final class WatchChannel {
    private let lock = NSLock()
    private var values: [Data] = []
    private var waiter: ModuleCompletion?
    private var closed = false

    func next(_ completion: @escaping ModuleCompletion) {
        lock.lock()
        if closed {
            lock.unlock()
            completion(.failure, Data("Observation is closed".utf8))
        } else if !values.isEmpty {
            let value = values.removeFirst()
            lock.unlock()
            completion(.success, value)
        } else if waiter != nil {
            lock.unlock()
            completion(.failure, Data("Observation already has a pending read".utf8))
        } else {
            waiter = completion
            lock.unlock()
        }
    }

    func offer(_ value: Data) {
        lock.lock()
        guard !closed else {
            lock.unlock()
            return
        }
        let callback = waiter
        if callback == nil {
            if values.count >= 4 { values.removeFirst() }
            values.append(value)
        } else {
            waiter = nil
        }
        lock.unlock()
        callback?(.success, value)
    }

    func close() {
        lock.lock()
        guard !closed else {
            lock.unlock()
            return
        }
        closed = true
        values.removeAll()
        let callback = waiter
        waiter = nil
        lock.unlock()
        callback?(.failure, Data("Observation stopped".utf8))
    }
}
